### DISCUSSION TITLE
[help] Read objects from dynamic branching?

### DISCUSSION THREAD

**Discussion by rkrug** [ropensci/targets#1299] [Help]
### Help

- [X] I understand and agree to <https://books.ropensci.org/targets/help.html>.

### Description

I have a pipeline which uses Static Branching

```
values <- data.frame(
    files = list.files(
        path = "input",
        pattern = "\\.csv$",
        full.names = FALSE
    ) |>
        gsub(pattern = "\\.csv$", replacement = "")

tar_map(
        values = values,
        ...
)
```

with the result that I  know the names of the objects to be used with `tar_read_raw()`. 

I use `tar_read_raw()` in a quarto document at the end which creates reports for each branch. Later, I will create reports which need all / multiple branches.

Now I am thinking about switching to dynamic branching, but how can I do the equivalent of  `tar_read_raw("object_BRANCH.NAME")`  (wnhat I did in the static branches)?

I know that I could define the report as a target, and I might look into using `tar_quarto_rep()`, but trying this, already cost me to much time, so that I will stuck with the existing (working) solution and put the "report as target" in the issue tracker.

**Comment by wlandau** [comment]
`tar_read()` and `tar_read_raw()` work with dynamic branches. Example:

```{r}
# _targets.R
library(targets)
list(
  tar_target(x, seq_len(2)),
  tar_target(y, x, pattern = map(x))
)
```

```{r}
tar_make()
#> ▶ dispatched target x
#> ● completed target x [0 seconds]
#> ▶ dispatched branch y_5015dfa5778ec11d
#> ● completed branch y_5015dfa5778ec11d [0 seconds]
#> ▶ dispatched branch y_5ea9d15a285adf66
#> ● completed branch y_5ea9d15a285adf66 [0 seconds]
#> ● completed pattern y
#> ▶ ended pipeline [0.125 seconds]
```

```{r}
tar_read_raw("y_5ea9d15a285adf66")
#> [1] 2
```

**Reply by rkrug** [reply]
Thanks. But my problem is, that the branches are based on file names, (the first input in the pipeline), and each report reads the objects based on a single filename. How can I get the part of the object name after the `_` (the bench name, I assume), from the file name? In static branching it is the file name, but in dynamic branching some hash?

**Reply by wlandau** [reply]
There isn't a clean way to identifying the specific upstream branch name from within a downstream branch, but you might not need to. If your pipeline begins with CSV files (as I think your original post has described) then you might be able to do something like this:

```r
list(
  tar_target(file_list, c("a.csv", "b.csv")),
  tar_target(file, file_list, pattern = map(file_list), format = "file"),
  tar_target(
    run_quarto,
    {
      html_file <- fs::path_ext_set(file, "html")
      quarto_render("source.qmd", output_file = html_file, execute_params = list(input = file))
      c("source.qmd", html_file)
    },
    pattern = map(file)
    format = "file"
  )
)
```

**Reply by ginolhac** [reply]
From another question by the same OP, this suggestion seems relevant: https://github.com/ropensci/targets/discussions/1297#discussioncomment-9752481
dynamic context, reading CSV files and full reprex

**Reply by rkrug** [reply]
Thanks a lot @wlandau and @ginolhac for your responses and patience. 

I will have to rest this project for some time but I will come back to this if necessary (the project will be necessary). 

I will also look a bit more into target and `quarto_render()` as I have the feeling the problem is that I am overlooking something (hopefully) trivial and if I get pent drops, everything will fall, into place.
